### ISSUE TITLE
EF Core Integration

### ISSUE THREAD

**Issue by diego81b** [gtaylor44/SqlBulkTools#40]
Hi @gtaylor44 
i write this extension method for integrating EntityFramework (in this case i'm working on .Net Core) into this library to reduce table configuration:

```
      public static BulkAddColumn<T> ForEFContext<T>(this BulkForCollection<T> bulk, DbContext context) {
            var entityType = context.Model.GetEntityTypes().FirstOrDefault(x => x.ClrType == typeof(T));
            BulkAddColumn<T> bulkColumn = null;
            if (entityType != null)
            {
                try
                {
                    var tableName = entityType.Relational().TableName;
                    var schema = entityType.Relational().Schema;
                    var tableConfig = bulk.WithTable(tableName).WithSchema(schema);

                    foreach (var propertyType in entityType.GetProperties())
                    {
                        var annotation =
                            propertyType.GetAnnotations().FirstOrDefault(p => p.Name.Equals("SqlServer:ColumnName"));
                        var columnName = annotation?.Value.ToString() ?? propertyType.Relational().ColumnName;
                        var lambda = GetPropertySelector<T>(propertyType.Name);
                        bulkColumn = bulkColumn == null
                            ? tableConfig.AddColumn(lambda, columnName)
                            : bulkColumn.AddColumn(lambda, columnName);
                    }
                    return bulkColumn;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
            throw new Exception("no type found");
     }
```

i don't know if should be useful or the right approach / goals for, but i'm glad to share with you

P.S.
Great library

Diego

**Comment by gtaylor44** [comment]
Thanks for sharing your code snippet. Could you share some benefits of this approach? Before and after examples?

Kind Regards,
Greg

**Comment by diego81b** [comment]
my goal is for no worry about managing table mapping, connection, transaction, etc... when we work with an ORM.

just to show you what i'm trying to do:

**Before**
```
using (var transaction = new TransactionScope())
            {
                using (var connection = new SqlConnection(connectionString))
                {
                    bulk.Setup<Company>()
                        .ForCollection(companies)
                        .WithTable("TB_COMPANY_BKP")
                        .WithSchema("DATI")
                        .AddColumn(x => x.Id, "ID_COMPANY")
                        .AddColumn(x => x.Name, "NAME")
                        .AddColumn(x => x.Email, "EMAIL")
                        .AddColumn(x => x.PhoneNumber, "PHONE_NUMBER")
                        .BulkInsert()
                        .SetIdentityColumn(x => x.Id)
                        .Commit(conn);
                }
            }
```

**After**
```
bulk.Setup<Company>()
                .ForCollection(companies)
                .ForEFContext(DbContext)
                .BulkInsert()
                .SetIdentityColumn(x => x.Id) // next step --> automatic identity column
                .Commit(true); // no worry about transaction and connection
```

i hope this help you
Diego

**Comment by gtaylor44** [comment]
Just some thoughts on the API changes..

1. Just looking at your comment 'next step --> automatic identity column'. There is currently an overload on SetIdentityColumn for outputting identity. If there is automatic identity detection, how will you be able to configure identity output? I've thought about checking if there is a [Key] or custom [Identity] attribute on the property and automatically setting it as an identity column in the past, but decided against it since it would be hard to have multiple configurations for the same model.

2. In the before example, The 4 calls to AddColumn can be replaced with AddAllColumns() which is a single line and states explicit intention. 

3. For selecting a table succinctly, in the latest version of the library, you can set schema in the 'WithTable' method. The before example
```
.WithTable("TB_COMPANY_BKP")
.WithSchema("DATI")
```

can be replaced with:
```
.WithTable("DATI.TB_COMPANY_BKP")
```

However, I do like the idea of being able to determine schema and table info from DbContext if EntityFramework is available. 

Do you think this could be refactored into an extension method of DbContext instead? If you can refactor a solution to extend DbContext I would happily merge a pull request. 

By the way, there have been some recent changes so make sure you have the latest copy of master.

**Comment by diego81b** [comment]
Thank you @gtaylor44 for your reply.

i'm not so expert (and i haven't a good english too) but i try to show you my point of view.
 "a change in the model mapping need a change on BulkTools configuration" with this on mind i'm studying a way to avoid this.

For extending DbContext what do you mean exactly? If u are saying to extend DbContext with Bulk functionality we need more than a simple extension method :smile:
The approach could be to create a derived class from DbContext and implement a bunch of "Bulk method" as the actual SaveChanges() do.

I like your idea man, as soon as possible, i'll take a look and i'll be happy to submit a PR (will be hard)

based on your consideration:
1. If i remember well there is a way in EF to recognize Primary Key without add attribute but i need more investigation.

2.  Does "AddAllColumns()" allow multiple custom naming ?
this is my situation:
**DB Table**
![table](https://cloud.githubusercontent.com/assets/9203744/24577001/10a06808-16c6-11e7-94b8-dda677ebe10d.JPG)
**Entity**
```
    public class Company : IEntity<int>
    {
        public int Id { get; set; }
        public ICollection<User> Users { get; set; }
        public ICollection<Project> Projects { get; set; }

        public string Name { get; set; }
        public string Email { get; set; }
        public string PhoneNumber { get; set; }

        public string ReferentName { get; set; }
        public string ReferentEmail { get; set; }
        public string ReferentPhoneNumber { get; set; }
    }
```

3. You are right

Thank you very much
Diego

**Comment by gtaylor44** [comment]
2. Yes you can do multiple custom column mappings with AddAllColumns() but in your example it doesn't have any benefit. The example shows that your naming convention differs between c# model and table. Every column needs to be mapped. Maybe an parameter on AddAllColumns() to convert case from PascalCase to UPPER_CASE would be beneficial for your usage. It's not something I've thought about before.

**Comment by gtaylor44** [comment]
And for extending the DbContext class... There's another library based on SqlBulkTools that you could look at. This leans more towards what you want to achieve I think and would be easier to integrate with.

[EntityFramework.BulkExtensions](https://github.com/tiagoln/EntityFramework.BulkExtensions)

**Comment by diego81b** [comment]
:+1:  Thank you very much @gtaylor44 for your support

**Comment by gtaylor44** [comment]
That's okay @diego81b . Just to recap... If you want to interact with DbContext, you're better off using the library I've linked to above. I did make some comments about a 'would-be-nice' DbContext hybrid feature but in truth, I think it's a bit over the top with little gains. 

To solve your main problem of '**reducing table configuration**', we could look at a way to convert casing for all columns so instead of mapping each column one by one, you could call AddAllColumns() once and set the case to UPPER_CASE. Your naming convention in SQL is consistent so it would work. 

Kind Regards,
Greg

**Comment by diego81b** [comment]
Yeah i was looking how it works  [EntityFramework.BulkExtensions](https://github.com/tiagoln/EntityFramework.BulkExtensions) but i think miss some functionalities (upsert, merge,  EF Core support, etc...) but never mind maybe i use it as inspiration of my own library.

About naming convention i agree with you that an upper_case func can be a plus.
Before work on this, keep in mind that will be many convention so instead of a upper-case converter should be very interesting planning one of this feature:

- String manipulator (for ex. a Func or Action as parameters)
- Regular Expression (RegExp as parameter)
- Build-In converter (for upper-case, prefix-suffix, PascalCase, camelCase, snake_case, kebab-case)

well done @gtaylor44 

Thanks
Diego

**Comment by borisdj** [comment]
I have made a class that encapsulates Bulks function with DbContext Extensions.
It enables you to call it like this:
````
context.BulkInsert(listOfEntities);
context.BulkUpdate(listOfEntities);
context.BulkDelete(listOfEntities);
````
It is in this issue:
[Extension of DbContext](https://github.com/gtaylor44/SqlBulkTools/issues/47)
